presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java

```diff
@@ -0,0 +1,560 @@
+                new PushDownDereferenceThroughProject(),
+                new PushDownDereferenceThroughUnnest(),
+                new PushDownDereferenceThroughSemiJoin(),
+                new PushDownDereferenceThroughJoin());
+    }
+
+    /**
+     * Extract dereferences and push them down to new ProjectNode below
+     * Transforms:
+     * <pre>
+     *  TargetNode(expression(a.x))
+     *  </pre>
+     * to:
+     * <pre>
+     *   ProjectNode(original symbols)
+     *    TargetNode(expression(symbol))
+     *      Project(symbol := a.x)
+     * </pre>
+     */
+    abstract class ExtractProjectDereferences<N extends PlanNode>
+            implements Rule<N>
+    {
+        private final Class<N> planNodeClass;
+
+        ExtractProjectDereferences(Class<N> planNodeClass)
+        {
+            this.planNodeClass = planNodeClass;
+        }
+
+        @Override
+        public Pattern<N> getPattern()
+        {
+            return Pattern.typeOf(planNodeClass);
+        }
+
+        @Override
+        public Result apply(N node, Captures captures, Context context)
+        {
+            Map<DereferenceExpression, VariableReferenceExpression> expressions =
+                    getDereferenceSymbolMap(ExpressionExtractor.extractExpressionsNonRecursive(node), context, metadata, sqlParser);
+
+            if (expressions.isEmpty()) {
+                return Result.empty();
+            }
+
+            return Result.ofPlanNode(new ProjectNode(context.getIdAllocator().getNextId(), rewrite(context, node, HashBiMap.create(expressions)), identityAssignmentsAsSymbolReferences(node.getOutputVariables())));
+        }
+
+        protected abstract N rewrite(Context context, N node, BiMap<DereferenceExpression, VariableReferenceExpression> expressions);
+    }
+
+    class ExtractFromFilter
+            extends ExtractProjectDereferences<FilterNode>
+    {
+        ExtractFromFilter()
+        {
+            super(FilterNode.class);
+        }
+
+        @Override
+        protected FilterNode rewrite(Context context, FilterNode node, BiMap<DereferenceExpression, VariableReferenceExpression> expressions)
+        {
+            PlanNode source = node.getSource();
+
+            ImmutableMap<VariableReferenceExpression, RowExpression>
+                    dereferencesMap =
+                    expressions.inverse().entrySet().stream().collect(toImmutableMap(Map.Entry::getKey,
+                            entry -> castToRowExpression(entry.getValue())));
+            Assignments assignments = Assignments.builder().putAll(identityAssignmentsAsSymbolReferences(source.getOutputVariables())).putAll(dereferencesMap).build();
+            ProjectNode projectNode = new ProjectNode(context.getIdAllocator().getNextId(), source, assignments);
+
+            RowExpression filter = castToRowExpression(ExpressionTreeRewriter.rewriteWith(new DereferenceReplacer(expressions), castToExpression(node.getPredicate())));
```

Comment:
this logic repeats in 3 places; consider extracting helper method:

```
    private RowExpression replaceDereferences(RowExpression rowExpression, Map<DereferenceExpression, VariableReferenceExpression> dereferences) {
        return castToRowExpression(rewriteWith(new DereferenceReplacer(dereferences), castToExpression(rowExpression)));
    }

            return new FilterNode(
                    context.getIdAllocator().getNextId(),
                    projectNode,
                    replaceDereferences(node.getPredicate(), expressions));
```

Reply:
fixed